Move assert.hpp from tt-metal API to ttstl

riverwuTT opened the pull request:
### Ticket

### Problem description
assert.hpp is used widely across meta, ttnn and other projects. Current directory structure suggests it belongs to the tt-metal API which is not appropriate for to describe it's use and implementation. 

### What's changed
Moved assert.hpp from tt-metal API to tt-stl.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17936087932)

riverwuTT committed:
move assert from tt-metal to ttstl

riverwuTT committed:
remove unused OStreamJoin

riverwuTT committed:
inline asserts and throws

riverwuTT committed:
move implementation details to detail namespace

ayerofieiev-tt reviewed (approved):
Thinking of our consumers. 
Does this change make them happier? 
Is it worth the "Replace All" hussle?

as we discussed offline, I want developers at TT be very mindful about it. 
We own our codebase and can make whatever changes we'd like, but I want each change to be an informed decision. This PR is technically correct and I am leaving an approve. But please adopt the customer perspective here too.

riverwuTT committed:
fix missing include

riverwuTT committed:
add #warning for original assert.hpp for soft transition

riverwuTT committed:
move more files to tt_stl/assert

riverwuTT committed:
update incorrect include

riverwuTT committed:
remove incorrect rebase

riverwuTT reviewed (commented):
Okay I manually checked all the updated include and marked out files that have more than a 2 line change.

riverwuTT committed:
fix extra edits to test_lightmetal.cpp

riverwuTT committed:
omit llrt extra include

riverwuTT committed:
omit drive-by for tt_metal.cpp

riverwuTT committed:
remove extra include

riverwuTT committed:
remove extra include

riverwuTT committed:
Remove drive-by formatting changes from llrt.cpp

Keep only the essential include path change from <assert.hpp> to <tt_stl/assert.hpp>

riverwuTT committed:
fix EOF

riverwuTT committed:
make linter happy

riverwuTT committed:
fix bad merge

riverwuTT committed:
Remove drive-by formatting changes from fabric_host_utils.cpp

Keep only the essential include path change from <tt-metalium/assert.hpp> to <tt_stl/assert.hpp>

riverwuTT committed:
Fix unintentional change in rtoptions.cpp

Keep only the essential include path change from "assert.hpp" to <tt_stl/assert.hpp>.
Remove unintentional UMD include path change.

riverwuTT committed:
removes assert.hpp from metal API

riverwuTT commented on tt_stl/tt_stl/assert.hpp:
```diff
@@ -0,0 +1,173 @@
+ */
+// NOLINTBEGIN(cppcoreguidelines-no-malloc)
+inline std::vector<std::string> backtrace(int size = 64, int skip = 1) {
+    std::vector<std::string> bt;
```
applied

riverwuTT commented on tt_stl/tt_stl/assert.hpp:
```diff
@@ -0,0 +1,173 @@
+ * @param[in] skip Skip the number of layers at the top of the stack
+ */
+// NOLINTBEGIN(cppcoreguidelines-no-malloc)
+inline std::vector<std::string> backtrace(int size = 64, int skip = 1) {
```
applied

riverwuTT commented on tt_stl/tt_stl/assert.hpp:
```diff
@@ -0,0 +1,173 @@
+inline std::string backtrace_to_string(int size = 64, int skip = 2, const std::string& prefix = "") {
+    std::vector<std::string> bt = backtrace(size, skip);
+    std::stringstream ss;
+    for (size_t i = 0; i < bt.size(); ++i) {
+        ss << prefix << bt[i] << std::endl;
+    }
```
applied

riverwuTT commented on tt_stl/tt_stl/assert.hpp:
```diff
@@ -0,0 +1,173 @@
+    char const* assert_type,
+    char const* condition_str,
+    fmt::format_string<Args const&...> fmt,
+    Args const&... args) {
+    tt_throw_impl(file, line, assert_type, condition_str, fmt, args...);
```
Good point.

riverwuTT commented on tt_stl/tt_stl/assert.hpp:
```diff
@@ -0,0 +1,173 @@
+    bool condition,
+    char const* condition_str,
+    fmt::format_string<Args const&...> fmt,
+    Args const&... args) {
```
Done

riverwuTT commented:
I'll do the perfect forwarding in another PR... something is being funny with fmt.

riverwuTT committed:
adopt some of the review suggestions

riverwuTT committed:
update new reference to old api/assert.hpp
